### ISSUE TITLE
Request: Dissective model class

### ISSUE THREAD

**Issue by JoKra1** [GWeindel/hmp#264]
Hey Gabriel 👋🏻,

Inspired by the implementation of the ``cumulative`` model, I had the idea for a ``dissective`` model class that also aims to iteratively identify HMP events.

Basically, the idea is to iteratively "dissect" the signal: take out the expected contribution of the latest event from every trial and time-point. Then continue searching for new events in the newly dissected signal. Repeat this until no events are found. This whole routine should probably be nested within an outer loop, since (I think, maybe you have a solution for this) the ``time_pars`` can only be updated after all events have been "dissected"..

Below some pseudo-code to illustrate the potential fitting routine:

```
fit_dissective(orig_data, template, max_events, max_iter):

  // Some initialisation
  channel_pars, time_pars = initialise_pars()
    
  for iter in 1:max_iter:
    
    // ``event_probs`` holds probability of event onset for up to max_events events
    event_probs = []
    // ``data`` holds iteratively dissected data
    data = copy(orig_data)
  
    for event in 1: max_events:
        
      // Fit single event model to current data
      event_model = EventModel(self.pattern, self.distribution, tolerance=self.tolerance,
                                                      n_events=1)
  
      event_model.fit(data, channel_pars[event], time_pars[event)
      
      // Update channel parameters from new fit
      channel_pars[event] = update_channel_pars(event_model)
      
      // Extract event_probs_new, holding probability of event onset for new event for every time-point and trial
      event_probs_new = event_model.event_probs
      
      // Compute the expected value of the (dissceted) EEG/PCA signal under current single event model.
      // This should just be the convolution of ``event_probs_new`` with ``self.pattern*channel_pars[event]``,
      // in any case this is a matrix of the same dimension of data, so we can dissect the current event out of data!
      E_data_new = compute_expected_signal(event_probs_new, self.pattern, channel_pars[event])
      
      // Dissect - i.e., iteratively remove expected impact of current event from data, so that data only holds only
      // information about new/missing events:
      data -= E_data_new
      
      // Finally, need to update event_probs. This is tricky - since the current event can be found anywhere
      // in-between previously found events. Might even differ on the trial level.. but for parameter updates
      // only the average onset is used, so I suggest sorting ``event_probs_new`` into ``event_probs``
      // based on that.
      event_probs = update_event_probs(event_probs_new)
      
      // And need some kind of detection whether likelihood change of finding new event is small enough to terminate
      if converged():
        break
  
    // In this iteration ``event`` events have been found - so ``event_probs`` holds as many onset probabilities so we can
    // update the ``time_pars`` for these events only
    time_pars = update_time_pars(event_probs, event)

    // Also need to perform some convergence check for the outer loop, probably based on change in and ``time_pars``
    if outer_converged():
      break

return channel_pars, time_pars
```

Phew, now that I have spelled it out, I do see some potential issues 😬:
- No guarantee that the order of events is the same at every iteration of the outer loop. I.e., first event identified on iteration 2 might not have been the first event identified at iteration 1. Theoretically, as parameter estimates become increasingly more refined this might not be an issue, but this could just generate loops in which the estimation gets stuck.
- No guarantee that the same number of events is found at every iteration of the outer loop. Might be a good idea to keep track of how often we find a certain number of events and then in the end just return only the number of events found most frequently.
- If there are repeated events with the same topography, then the dissection step should take them all out at once. Now, this might actually be a feature rather than a bug but I am not sure.
- Naturally, the algorithm is very heuristic: If you were to dissect events in a different order you might find a different set of total events. But this is a similar problem for the ``cumulative`` method anyway.
- No overfit protection (again just like the ``cumulative`` method): at no point is there a check whether the identified number of events generalises beyond the data used during fitting.

The big advantage: there is no need to worry about overlapping events, since every new dissection is performed conditional on the expected contribution of previously identified events! So this is one option to avoid the problem outlined in #262!

**Comment by GWeindel** [comment]
I think your approach could indeed address issue #262. It really resembles the RIDE algorithm (https://eeg-ride.readthedocs.io/en/latest/, see also https://pmc.ncbi.nlm.nih.gov/articles/PMC6815663/) except that in your case you don't specify the number of component nor whether they're tied to an external event. But in the current implementation of HMP I don't see yet an easy way to implement it without it being hacky.

But I wonder whether overlap is really the problem. Event width of underlying events have to vary and probably by a large amount (e.g. CPP/LRP vs C1). That makes me think of another approach, and, to stick with the -ive suffix we could call it `agglomerative`.
The idea would be that events are detected with a very short event width (e.g. 10ms) and events with the same topography and short time distributions get lumped together into a single larger event whos width is given by the number of agglomerated events. I would need to think about how this would be implemented in the current likelihood computation.
